# Fase 1 · Análisis Exploratorio de Datos (EDA)
 
## Objetivo
 
El objetivo de esta fase es realizar una exploración inicial del dataset **DataCo Smart Supply Chain**, el cual simula las operaciones comerciales, logísticas y de cadena de suministro de una organización global. Esta revisión permite comprender la estructura multidimensional de los datos, evaluar su calidad y detectar posibles inconsistencias o sesgos que puedan afectar a las fases posteriores de modelado predictivo y analítica avanzada.
 
En esta etapa nos centramos en:
 
- **Identificar las dimensiones operativas:** Clasificar y entender las variables comerciales (ventas, clientes), logísticas (modos de envío, tiempos de entrega) y de seguridad (estados de órdenes y alertas de fraude).
- **Analizar la consistencia estructural:** Evaluar las dimensiones del dataset, los tipos de datos (fechas, variables categóricas, numéricas) y la coherencia interna entre registros relacionados.
- **Evaluar la calidad de los datos:** Detectar la presencia de valores nulos, registros duplicados, inconsistencias en las coordenadas geográficas y datos potencialmente sensibles o expuestos.
- **Explorar variables críticas del negocio:** Analizar el comportamiento de métricas clave como los días de envío reales frente a los estimados (`SLA`), los patrones de fraude en los métodos de pago y la distribución de ventas por regiones.
- **Anticipar necesidades de ingeniería de características:** Identificar requerimientos específicos de limpieza, normalización de columnas, transformación de variables temporales y preparación de datos para la fase ETL.
 
Este análisis proporciona una base sólida para definir la estrategia de preparación de datos, asegurando un pipeline robusto para la toma de decisiones estratégicas en la cadena de suministro.

In [36]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
import warnings
import sys
import os # Esto sube un nivel en las carpetas para encontrar la raíz del proyecto
sys.path.append(os.path.abspath(os.path.join('..')))
import src.exploration as e

In [37]:
warnings.filterwarnings("ignore")
pd.set_option('display.max_columns', None) # para poder visualizar todas las columnas de los DataFrames

In [38]:
df_data = pd.read_csv("../files/raw/dataco_supply_chain.csv", sep=None, engine="python", encoding="latin-1")

In [39]:
e.info_df(df_data)


INFORME TÉCNICO DEL DATAFRAME

🔹 Dimensiones:
- Filas: 180519
- Columnas: 53

🔹 Info general:
<class 'pandas.DataFrame'>
RangeIndex: 180519 entries, 0 to 180518
Data columns (total 53 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   Type                           180519 non-null  str    
 1   Days for shipping (real)       180519 non-null  int64  
 2   Days for shipment (scheduled)  180519 non-null  int64  
 3   Benefit per order              180519 non-null  float64
 4   Sales per customer             180519 non-null  float64
 5   Delivery Status                180519 non-null  str    
 6   Late_delivery_risk             180519 non-null  int64  
 7   Category Id                    180519 non-null  int64  
 8   Category Name                  180519 non-null  str    
 9   Customer City                  180519 non-null  str    
 10  Customer Country               180519 non-null  str    
 11  Custome

,count,unique,top,freq
Type,180519,4,DEBIT,69295
Delivery Status,180519,4,Late delivery,98977
Category Name,180519,50,Cleats,24551
Customer City,180519,563,Caguas,66770
Customer Country,180519,2,EE. UU.,111146
Customer Email,180519,1,XXXXXXXXX,180519
Customer Fname,180519,782,Mary,65150
Customer Lname,180511,1109,Smith,64104
Customer Password,180519,1,XXXXXXXXX,180519
Customer Segment,180519,3,Consumer,93504



🔹 Descriptivo variables cuantitativas:


,count,mean,std,min,25%,50%,75%,max
Days for shipping (real),180519.0,3.50,1.62,0.00,2.00,3.00,5.00,6.00
Days for shipment (scheduled),180519.0,2.93,1.37,0.00,2.00,4.00,4.00,4.00
Benefit per order,180519.0,21.97,104.43,-4274.98,7.00,31.52,64.80,911.80
Sales per customer,180519.0,183.11,120.04,7.49,104.38,163.99,247.40,1939.99
Late_delivery_risk,180519.0,0.55,0.50,0.00,0.00,1.00,1.00,1.00
Category Id,180519.0,31.85,15.64,2.00,18.00,29.00,45.00,76.00
Customer Id,180519.0,6691.38,4162.92,1.00,3258.50,6457.00,9779.00,20757.00
Customer Zipcode,180516.0,35921.13,37542.46,603.00,725.00,19380.00,78207.00,99205.00
Department Id,180519.0,5.44,1.63,2.00,4.00,5.00,7.00,12.00
Latitude,180519.0,29.72,9.81,-33.94,18.27,33.14,39.28,48.78



🔹 Muestra del dataframe:


,Type,Days for shipping (real),Days for shipment (scheduled),Benefit per order,Sales per customer,Delivery Status,Late_delivery_risk,Category Id,Category Name,Customer City,Customer Country,Customer Email,Customer Fname,Customer Id,Customer Lname,Customer Password,Customer Segment,Customer State,Customer Street,Customer Zipcode,Department Id,Department Name,Latitude,Longitude,Market,Order City,Order Country,Order Customer Id,order date (DateOrders),Order Id,Order Item Cardprod Id,Order Item Discount,Order Item Discount Rate,Order Item Id,Order Item Product Price,Order Item Profit Ratio,Order Item Quantity,Sales,Order Item Total,Order Profit Per Order,Order Region,Order State,Order Status,Order Zipcode,Product Card Id,Product Category Id,Product Description,Product Image,Product Name,Product Price,Product Status,shipping date (DateOrders),Shipping Mode
123925,TRANSFER,3,4,41.740002,123.489998,Advance shipping,0,18,Men's Footwear,Caguas,Puerto Rico,XXXXXXXXX,Mary,5612,Huang,XXXXXXXXX,Corporate,PR,7417 Honey Mountain Lookout,725.0,4,Apparel,18.296637,-66.370544,LATAM,San Pablo de las Salinas,México,5612,5/8/2017 22:37,58842,403,6.5,0.05,147281,129.990005,0.34,1,129.990005,123.489998,41.740002,Central America,México,PROCESSING,NaN,403,18,NaN,http://images.acmesports.sports/Nike+Men%27s+C...,Nike Men's CJ Elite 2 TD Football Cleat,129.990005,0,5/11/2017 22:37,Standard Class
133650,TRANSFER,6,4,88.000000,351.980011,Late delivery,1,45,Fishing,San Jose,EE. UU.,XXXXXXXXX,Susan,1298,Fuller,XXXXXXXXX,Consumer,CA,421 Fallen Byway,95125.0,7,Fan Shop,37.292233,-121.881279,LATAM,Juiz de Fora,Brasil,1298,3/16/2015 3:24,5080,1004,48.0,0.12,12693,399.980011,0.25,1,399.980011,351.980011,88.000000,South America,Minas Gerais,PROCESSING,NaN,1004,45,NaN,http://images.acmesports.sports/Field+%26+Stre...,Field & Stream Sportsman 16 Gun Fire Safe,399.980011,0,3/22/2015 3:24,Standard Class
41193,DEBIT,2,4,141.949997,363.980011,Advance shipping,0,45,Fishing,Caguas,Puerto Rico,XXXXXXXXX,Helen,5600,Smith,XXXXXXXXX,Corporate,PR,2736 Blue Crest,725.0,7,Fan Shop,18.292276,-66.370613,LATAM,Mejicanos,El Salvador,5600,4/24/2015 20:41,7801,1004,36.0,0.09,19514,399.980011,0.39,1,399.980011,363.980011,141.949997,Central America,San Salvador,COMPLETE,NaN,1004,45,NaN,http://images.acmesports.sports/Field+%26+Stre...,Field & Stream Sportsman 16 Gun Fire Safe,399.980011,0,4/26/2015 20:41,Standard Class
148293,DEBIT,2,2,13.390000,212.500000,Shipping on time,0,24,Women's Apparel,Caguas,Puerto Rico,XXXXXXXXX,Mary,2824,Bray,XXXXXXXXX,Home Office,PR,18 Jagged Willow Beach,725.0,5,Golf,18.237143,-66.370598,LATAM,Saltillo,México,2824,3/25/2017 9:43,55791,502,37.5,0.15,139522,50.000000,0.06,5,250.000000,212.500000,13.390000,Central America,Coahuila,COMPLETE,NaN,502,24,NaN,http://images.acmesports.sports/Nike+Men%27s+D...,Nike Men's Dri-FIT Victory Golf Polo,50.000000,0,3/27/2017 9:43,Second Class
125228,PAYMENT,2,2,83.150002,245.990005,Shipping on time,0,17,Cleats,Caguas,Puerto Rico,XXXXXXXXX,Mary,7972,Hoffman,XXXXXXXXX,Consumer,PR,5546 Old Rabbit Towers,725.0,4,Apparel,18.295347,-66.370636,Europe,Frankfurt,Alemania,7972,9/19/2017 6:40,67976,364,54.0,0.18,169939,299.989990,0.34,1,299.989990,245.990005,83.150002,Western Europe,Hesse,PENDING_PAYMENT,NaN,364,17,NaN,http://images.acmesports.sports/Total+Gym+1400,Total Gym 1400,299.989990,0,9/21/2017 6:40,Second Class


In [40]:
df_data.head()

,Type,Days for shipping (real),Days for shipment (scheduled),Benefit per order,Sales per customer,Delivery Status,Late_delivery_risk,Category Id,Category Name,Customer City,Customer Country,Customer Email,Customer Fname,Customer Id,Customer Lname,Customer Password,Customer Segment,Customer State,Customer Street,Customer Zipcode,Department Id,Department Name,Latitude,Longitude,Market,Order City,Order Country,Order Customer Id,order date (DateOrders),Order Id,Order Item Cardprod Id,Order Item Discount,Order Item Discount Rate,Order Item Id,Order Item Product Price,Order Item Profit Ratio,Order Item Quantity,Sales,Order Item Total,Order Profit Per Order,Order Region,Order State,Order Status,Order Zipcode,Product Card Id,Product Category Id,Product Description,Product Image,Product Name,Product Price,Product Status,shipping date (DateOrders),Shipping Mode
0,DEBIT,3,4,91.250000,314.640015,Advance shipping,0,73,Sporting Goods,Caguas,Puerto Rico,XXXXXXXXX,Cally,20755,Holloway,XXXXXXXXX,Consumer,PR,5365 Noble Nectar Island,725.0,2,Fitness,18.251453,-66.037056,Pacific Asia,Bekasi,Indonesia,20755,1/31/2018 22:56,77202,1360,13.110000,0.04,180517,327.75,0.29,1,327.75,314.640015,91.250000,Southeast Asia,Java Occidental,COMPLETE,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,2/3/2018 22:56,Standard Class
1,TRANSFER,5,4,-249.089996,311.359985,Late delivery,1,73,Sporting Goods,Caguas,Puerto Rico,XXXXXXXXX,Irene,19492,Luna,XXXXXXXXX,Consumer,PR,2679 Rustic Loop,725.0,2,Fitness,18.279451,-66.037064,Pacific Asia,Bikaner,India,19492,1/13/2018 12:27,75939,1360,16.389999,0.05,179254,327.75,-0.80,1,327.75,311.359985,-249.089996,South Asia,Rajastán,PENDING,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/18/2018 12:27,Standard Class
2,CASH,4,4,-247.779999,309.720001,Shipping on time,0,73,Sporting Goods,San Jose,EE. UU.,XXXXXXXXX,Gillian,19491,Maldonado,XXXXXXXXX,Consumer,CA,8510 Round Bear Gate,95125.0,2,Fitness,37.292233,-121.881279,Pacific Asia,Bikaner,India,19491,1/13/2018 12:06,75938,1360,18.030001,0.06,179253,327.75,-0.80,1,327.75,309.720001,-247.779999,South Asia,Rajastán,CLOSED,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/17/2018 12:06,Standard Class
3,DEBIT,3,4,22.860001,304.809998,Advance shipping,0,73,Sporting Goods,Los Angeles,EE. UU.,XXXXXXXXX,Tana,19490,Tate,XXXXXXXXX,Home Office,CA,3200 Amber Bend,90027.0,2,Fitness,34.125946,-118.291016,Pacific Asia,Townsville,Australia,19490,1/13/2018 11:45,75937,1360,22.940001,0.07,179252,327.75,0.08,1,327.75,304.809998,22.860001,Oceania,Queensland,COMPLETE,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/16/2018 11:45,Standard Class
4,PAYMENT,2,4,134.210007,298.250000,Advance shipping,0,73,Sporting Goods,Caguas,Puerto Rico,XXXXXXXXX,Orli,19489,Hendricks,XXXXXXXXX,Corporate,PR,8671 Iron Anchor Corners,725.0,2,Fitness,18.253769,-66.037048,Pacific Asia,Townsville,Australia,19489,1/13/2018 11:24,75936,1360,29.500000,0.09,179251,327.75,0.45,1,327.75,298.250000,134.210007,Oceania,Queensland,PENDING_PAYMENT,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/15/2018 11:24,Standard Class


In [53]:
df_data["shipping date (DateOrders)"].value_counts()

shipping date (DateOrders)
5/9/2015 18:02      10
9/27/2015 20:28     10
9/24/2015 23:48     10
8/1/2015 18:37      10
11/15/2016 13:45    10
                    ..
1/27/2016 18:59      1
1/28/2016 19:25      1
1/26/2016 6:49       1
1/23/2016 7:10       1
1/21/2016 5:56       1
Name: count, Length: 63701, dtype: int64

In [41]:
df_data["Shipping Mode"].value_counts()

Shipping Mode
Standard Class    107752
Second Class       35216
First Class        27814
Same Day            9737
Name: count, dtype: int64

In [42]:
df_data["Customer Country"].value_counts()

Customer Country
EE. UU.        111146
Puerto Rico     69373
Name: count, dtype: int64

In [43]:
df_data["Customer State"].unique()

<StringArray>
[   'PR',    'CA',    'NY',    'FL',    'MA',    'IL',    'MT',    'PA',
    'MI',    'TX',    'DE',    'GA',    'MD',    'OH',    'HI',    'NJ',
    'WI',    'AZ',    'CO',    'MN',    'NC',    'NM',    'OR',    'SC',
    'VA',    'UT',    'WA',    'KY',    'WV',    'RI',    'CT',    'LA',
    'TN',    'DC',    'ND',    'MO',    'IN',    'ID',    'NV',    'KS',
    'AR',    'OK',    'AL',    'IA', '95758', '91732']
Length: 46, dtype: str

In [44]:
df_data["Order Status"].unique()

<StringArray>
[       'COMPLETE',         'PENDING',          'CLOSED', 'PENDING_PAYMENT',
        'CANCELED',      'PROCESSING', 'SUSPECTED_FRAUD',         'ON_HOLD',
  'PAYMENT_REVIEW']
Length: 9, dtype: str

In [45]:
df_data["Order Region"].unique()

<StringArray>
[ 'Southeast Asia',      'South Asia',         'Oceania',    'Eastern Asia',
       'West Asia',    'West of USA ',      'US Center ',     'West Africa',
  'Central Africa',    'North Africa',  'Western Europe', 'Northern Europe',
 'Central America',       'Caribbean',   'South America',     'East Africa',
 'Southern Europe',     'East of USA',          'Canada', 'Southern Africa',
    'Central Asia',  'Eastern Europe',  'South of  USA ']
Length: 23, dtype: str

In [46]:
df_data["Market"].unique()

<StringArray>
['Pacific Asia', 'USCA', 'Africa', 'Europe', 'LATAM']
Length: 5, dtype: str

In [47]:
df_data["Category Name"].unique()

<StringArray>
[      'Sporting Goods',               'Cleats',        'Shop By Sport',
      'Women's Apparel',          'Electronics',         'Boxing & MMA',
     'Cardio Equipment',             'Trade-In',     'Kids' Golf Clubs',
   'Hunting & Shooting',  'Baseball & Softball',       'Men's Footwear',
     'Camping & Hiking', 'Consumer Electronics',             'Cameras ',
            'Computers',           'Basketball',               'Soccer',
       'Girls' Apparel',          'Accessories',     'Women's Clothing',
               'Crafts',       'Men's Clothing',     'Tennis & Racquet',
  'Fitness Accessories',      'As Seen on  TV!',           'Golf Balls',
    'Strength Training',  'Children's Clothing',             'Lacrosse',
                'Baby ',              'Fishing',               'Books ',
                 'DVDs',                 'CDs ',               'Garden',
               'Hockey',         'Pet Supplies',    'Health and Beauty',
                'Music',          'Vi

In [48]:
df_data["Customer Segment"].unique()

<StringArray>
['Consumer', 'Home Office', 'Corporate']
Length: 3, dtype: str

In [49]:
df_data.groupby("Customer State")["Order Zipcode"].unique()

Customer State
91732                                                [nan]
95758                                                [nan]
AL                [nan, 60462.0, 98105.0, 7960.0, 22980.0]
AR               [nan, 43130.0, 90032.0, 39212.0, 43302.0]
AZ       [nan, 10035.0, 32216.0, 43229.0, 46368.0, 8001...
CA       [nan, 95123.0, 19134.0, 31907.0, 77041.0, 2230...
CO       [nan, 26003.0, 38109.0, 92024.0, 19134.0, 1002...
CT       [nan, 89115.0, 19140.0, 95051.0, 90049.0, 7501...
DC       [nan, 73120.0, 31907.0, 48227.0, 47201.0, 9203...
DE       [nan, 43055.0, 60505.0, 90049.0, 90032.0, 7050...
FL       [nan, 38109.0, 72401.0, 60610.0, 6708.0, 43017...
GA       [nan, 10035.0, 97206.0, 98661.0, 87124.0, 7570...
HI       [60016.0, nan, 75034.0, 13440.0, 89031.0, 1000...
IA               [nan, 10011.0, 85023.0, 48911.0, 78041.0]
ID               [nan, 38301.0, 48234.0, 20735.0, 19711.0]
IL       [nan, 13021.0, 79762.0, 32216.0, 49201.0, 9811...
IN       [nan, 37042.0, 60623.0, 94110.0,

In [50]:
df_token = pd.read_csv("../files/raw/tokenized_access_logs.csv")

In [51]:
e.info_df(df_token)


INFORME TÉCNICO DEL DATAFRAME

🔹 Dimensiones:
- Filas: 469977
- Columnas: 8

🔹 Info general:
<class 'pandas.DataFrame'>
RangeIndex: 469977 entries, 0 to 469976
Data columns (total 8 columns):
 #   Column      Non-Null Count   Dtype
---  ------      --------------   -----
 0   Product     469977 non-null  str  
 1   Category    469977 non-null  str  
 2   Date        469977 non-null  str  
 3   Month       469977 non-null  str  
 4   Hour        469977 non-null  int64
 5   Department  469977 non-null  str  
 6   ip          469977 non-null  str  
 7   url         469977 non-null  str  
dtypes: int64(1), str(7)
memory usage: 28.7 MB

🔹 Nulos:
Product       0.0
Category      0.0
Date          0.0
Month         0.0
Hour          0.0
Department    0.0
ip            0.0
url           0.0
dtype: float64

🔹 Duplicados:
Total duplicados: 3249

🔹 Descriptivo variables cualitativas:


,count,unique,top,freq
Product,469977,76,Perfect Fitness Perfect Rip Deck,27878
Category,469977,33,cleats,27878
Date,469977,160815,9/14/2017 20:34,156
Month,469977,5,Sep,137238
Department,469977,6,outdoors,79926
ip,469977,3340,157.21.93.193,566
url,469977,152,/department/apparel/category/cleats/product/Pe...,20258



🔹 Descriptivo variables cuantitativas:


,count,mean,std,min,25%,50%,75%,max
Hour,469977.0,14.59,5.57,0.0,10.0,15.0,20.0,23.0



🔹 Muestra del dataframe:


,Product,Category,Date,Month,Hour,Department,ip,url
162813,Nike Men's Free 5.0+ Running Shoe,cardio equipment,11/7/2017 9:05,Nov,9,footwear,40.137.4.17,/department/footwear/category/cardio%20equipme...
124416,Yakima DoubleDown Ace Hitch Mount 4-Bike Rack,strength training,9/11/2017 13:46,Sep,13,footwear,190.114.48.134,/department/footwear/category/strength%20train...
200261,adidas Kids' RG III Mid Football Cleat,featured shops,12/23/2017 18:47,Dec,18,apparel,57.41.66.112,/department/apparel/category/featured%20shops/...
243583,adidas Youth Germany Black/Red Away Match Soc,girls' apparel,11/19/2017 13:14,Nov,13,golf,138.241.223.150,/department/golf/category/girls'%20apparel/pro...
253364,Nike Men's Free 5.0+ Running Shoe,cardio equipment,12/30/2017 9:24,Dec,9,footwear,21.161.97.90,/department/footwear/category/cardio%20equipme...


In [52]:
df_token["Month"].unique()

<StringArray>
['Sep', 'Oct', 'Dec', 'Jan', 'Nov']
Length: 5, dtype: str

# Conclusiones finales del EDA · DataCo
 
- **Calidad de Datos y Gestión de Nulos:** La integridad general del dataset es excelente en la mayoría de sus variables transaccionales y financieras. La excepción principal se concentra en la columna `Order Zipcode`, la cual presenta valores nulos que requieren una estrategia de imputación o limpieza en la fase ETL, dado que pueden afectar el análisis geoespacial del destino de las órdenes.
 
- **Duplicados y Estructura Lógica:** Las repeticiones detectadas en los registros no constituyen errores de duplicidad aleatoria; se justifican plenamente por la estructura lógica y granular del dataset, donde un mismo número de orden (`Order Id`) puede contener múltiples ítems de productos individuales o registrar distintas actualizaciones en el historial de envío.
 
- **Políticas de Privacidad y Reducción de Dimensiones (Eliminación):** Con el fin de optimizar el rendimiento del almacenamiento en el ecosistema de Big Data y garantizar el cumplimiento estricto de las normativas de protección de datos (debido a la naturaleza sintética/expuesta de cierta información), se determina la eliminación definitiva de las siguientes columnas no relevantes o sensibles:
  - `Customer Email` (Privacidad)
  - `Customer Password` (Seguridad)
  - `Product Description` (Columna redundante o vacía en el catálogo de productos)
  - `Order State` (Duplicidad con las métricas regionales o de envío)
 
- **Estandarización y Corrección de Anomalías:** Se detectaron inconsistencias críticas en la columna de ubicación del cliente (`Customer State`). Mientras que la mayoría de los registros adoptan el formato estándar de código de estado de dos letras (ej. NY, TX), se identificaron códigos postales numéricos (como *95758* y *91732*) en dicho campo. Como acción obligatoria de limpieza, se plantea mapear y reemplazar estos números por el estado correcto al que pertenecen (`CA`).
 
- **Optimización de Estructura y Formato de Columnas:** Dado que las columnas ya se encuentran en el idioma inglés, la transformación se centrará en estandarizar su formato de texto. Se aplicará una conversión estricta a minúsculas y se reemplazarán los espacios por guiones bajos (formato `snake_case`) para homogeneizar los nombres de las variables (ej. de `Order Id` a `order_id`). Asimismo, se transformarán las variables temporales desde texto plano a formato `datetime` nativo.
 
- **Integración y Complementariedad:** Las tablas analizadas demuestran ser altamente complementarias para construir un modelo en estrella o copo de nieve, permitiendo cruzar de manera directa las dimensiones de Clientes, Productos y Geografía con la tabla de hechos de Órdenes y Envíos.
 
---
 
Esta fase concluye con éxito, dejando una hoja de ruta clara, justificada y accionable para la siguiente etapa del pipeline: **Limpieza, Transformación y Carga (ETL)**.